### PR TITLE
Added mode for raytracing fluids for all gadgets, and mode for placing in or on blocks for the building gadget.

### LINKED ISSUE #243: Allow Placing on Liquid's
As per rquest of the Discord chat: Allow placing on Liquid's. Aka, allow the Building/Exchanging Gadget to build directly on liquid's like water. Should be toggelable similiar to the connected Surface mode.

3 states were suggested:
- ignored = the current way of doing it
- inLiquid = replace the topmost liquid block 
- onLiquid = treat Liquids as solid blocks, aka place on top of them (building/copy-paste only)

### PR DESCRIPTION
**Pull request by Phylogeny**
Resolves #243 and supersedes #247.
Given the mix of placing in blocks vs on blocks in the building gadget, the ability to toggle that for everything currently only on top is more useful than tying it to fluid raytracing.

### CODE REVIEW

File: src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetBuilding.java [hunk 1]
```diff
@@ -115,6 +116,16 @@ public static ToolMode getToolMode(ItemStack stack) {
         return mode;
     }
 
+    public static boolean shouldPlaceAtop(ItemStack stack) {
+        return !NBTHelper.getOrNewTag(stack).getBoolean("start_inside");
```
Comment (MajorTuvok) [inline src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetBuilding.java]: Could you convert this to a constant please?

File: src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetGeneric.java [hunk 2]
```diff
@@ -150,6 +150,19 @@ public static void toggleConnectedArea(EntityPlayer player, ItemStack stack) {
         player.sendStatusMessage(new TextComponentString(TextFormatting.AQUA + new TextComponentTranslation("message.gadget.connected" + suffix).getUnformattedComponentText() + ": " + getConnectedArea(stack)), true);
     }
 
+    public static boolean shouldRayTraceFluid(ItemStack stack) {
+        return NBTHelper.getOrNewTag(stack).getBoolean("raytrace_fluid");
```
Comment (MajorTuvok) [inline src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetGeneric.java]: Could you convert this to a constant please?

File: src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetBuilding.java (hunk 1 above)
Comment (Phylogeny) [reply src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetBuilding.java]: I think that should be done in a dedicated push or PR. The code base is full of these, and there should be a coordinated effort to handle them consistently, at some point.

Comment (MajorTuvok) [reply src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetBuilding.java]: ok

**Comment by MajorTuvok** [comment]
As we discussed on discord: creating these 2 classes is out of the scope of this PR...

**Pull request merged**